### DISCUSSION TITLE
Can array columns be dumped with Python types instead of Numpy when using `.to_df()`?

### DISCUSSION THREAD

**Discussion by fjossandon** [duckdb/duckdb#14451] [Q&A] [answered]
Hello,
I have the following problem. I created a table in DuckDB that has an array of integers column (`INTEGER[]`). When I dump the content to a dataframe with `.to_df()`, the array column in the dataframe has Numpy types:
```
type(data["array_col"][0])
<class 'numpy.ndarray'>

type(data["array_col"][0][0])
<class 'numpy.int32'>
```
The cell value is "numpy.ndarray", and the integers in the array are numpy.int32... This causes problems later when I try to insert the dataframe content into a Postgres table because psycopg2 cannot handle it:
```
psycopg2.ProgrammingError: can't adapt type 'numpy.ndarray'
```
Looking for solutions, I found that I may have to iterate somehow over the column using `.tolist()` to convert the types to built-in Python types (list and ints) (https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html). In principle, my function receives the table name to retrieve it from DuckDB and returns the dataframe... but it doesn't know the table column types, so I would need to check the type of all columns every time to find if any has arrays and then reformat it accordingly before returning it.

My question is if maybe there is an optional argument that can be used in `.to_df()` that generates the dataframe using Python types instead of Numpy types, or if there is a simple way to achieve this.

Best regards,

**Comment by soerenwolfers** [comment, marked as answer]
You can do 

```
array_cols = df.select_dtypes(include=[np.ndarray]).columns
df[array_cols] = df[array_cols].applymap(np.ndarray.tolist)
```

Also, numpy arrays are ubiquitous in Python, they are almost always more efficient than Python lists, and
np.int32 is more faithful to duckdb's fixed size integer data type than Python integers
 -- you should open a pyscopg2 ticket to support numpy arrays  ;)

**Reply by fjossandon** [reply]
Hello @soerenwolfers,
Thanks for the advice, your solution worked perfectly, although I got a warning about using `map` instead of `applymap`:
```
<input>:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
```
So I settled for this:
```
    array_cols = table_data.select_dtypes(include=[np.ndarray]).columns
    if not array_cols.empty:
        table_data[array_cols] = table_data[array_cols].map(np.ndarray.tolist)
```
Regarding pyscopg2 supporting Numpy arrays, I found this closed issue:
https://github.com/psycopg/psycopg2/issues/1718

> In Numpy 2 it looks like they changed the repr() of these types, as well as the MRO. float64 fails for the different repr(), float32 fails because it doesn't inherit from float.
> Numpy 1.x support was only accidental and not tested; it worked because these types had both repr and MRO compatible with Python floats. Therefore, registering an adapter is the only supported way. Maybe, in your adapter, return AsIs(str(numpy_float)) would be better.
> Psycopg 3.2 has official support for numpy scalars and it is tested for both numpy version <2 and 2. We are not planning to add builtin support in psycopg2.

So it seems that there is only partial support for Numpy types, but its not their priority.
